### PR TITLE
8341901: Using 'var' keyword switch pattern matching causes compiler error

### PR DESCRIPTION
**Pull request by lahodaj**
Consider code like:
```
public class T {
   record R(N.I i) {}
   int test(Object o) {
       return switch (o) {
           case R(var nested) -> 0;
           default -> 0;
       };
   }
   static class N<T> {
       interface I {}
   }
}
```

This fails to compile since JDK 23, due to:
```
$ javac T.java 
error: cannot select a static class from a parameterized type
1 error
```

The reason for the error is this: the type of `nested` is inferred to `T.N.I`. This is correct. javac will then construct a synthetic AST for it, and the AST will be structurally correct as well: `T.N.I`. But a) the `Type` attached to `T.N` will be `T.N<T>` (which by itself is not correct), and b) after the synthetic AST is created, `Check.validate` is called on the type's AST, and fails, as the types is sees correspond to `T.N<T>.I`, which is illegal.

Note the synthetic AST is also set for local variable type inference, but the `validate` is called *before* the synthetic AST is created.

This PR proposes to do two things:
- move the `validate` call before the synthetic AST creation for `visitBindingPattern`, to mimic the behavior for `var`s.
- the `TreeMaker` is tweaked to inject erased types instead of parameterized types when generating qualified identifiers for classes or interfaces. This should correspond more closely to what happens when one types `T.N.I` in the source code.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341901](https://bugs.openjdk.org/browse/JDK-8341901): Using 'var' keyword switch pattern matching causes compiler error (**Bug** - P3)

### Reviewers
 * [Aggelos Biboudis](https://openjdk.org/census#abimpoudis) (@biboudis - **Reviewer**) Review applies to [7bf21b9c](https://git.openjdk.org/jdk/pull/21495/files/7bf21b9cf95912c9c0ff7ab10461e5d49d16f7b4)
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21495/head:pull/21495` \
`$ git checkout pull/21495`

Update a local copy of the PR: \
`$ git checkout pull/21495` \
`$ git pull https://git.openjdk.org/jdk.git pull/21495/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21495`

View PR using the GUI difftool: \
`$ git pr show -t 21495`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21495.diff">https://git.openjdk.org/jdk/pull/21495.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21495#issuecomment-2411369592)
</details>

### CODE REVIEW

**Review by biboudis — APPROVED** [review]
According to my understanding, this bug is fixed by adjusting the `TreeMaker` to use the erased types. As a result the path `T.N.I ` is correct and the tests pass, correctly. Moving `chk.validate` is an optimization right? In this case `tree.var.vartype` will be null for both tests so the validate does nothing.

Otherwise LGTM

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeMaker.java [hunk 1]
```diff
@@ -741,9 +741,15 @@ public JCFieldAccess Select(JCExpression base, Symbol sym) {
      *  to make the reference unique.
      */
     public JCExpression QualIdent(Symbol sym) {
-        return isUnqualifiable(sym)
+        JCExpression result = isUnqualifiable(sym)
             ? Ident(sym)
             : Select(QualIdent(sym.owner), sym);
+
+        if (sym.kind == TYP) {
+            result.setType(types.erasure(sym.type));
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeMaker.java:749]: can we do this outside of TreeMaker? I mean TreeMaker could be used to generate trees for a phase for which erasure hasn't occurred yet. It should be the caller's option to decide if the obtained tree should be erased or not

Comment (lahodaj) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeMaker.java:749]: Doing this outside of TreeMaker would be fairly difficult (as that would mean the caller would need to traverse the created tree and clean the types).

This change is not strictly necessary to solve the problem - moving the `chk.validate` is enough. But, it seems to me that type attributes in the AST that are produced by the `TreeMaker.QualIdent` do not make much sense. Specifically, passing the ClassSymbol for `java.util.List` here will produce a tree whose `.type` will be `java.util.List<E>`, where `E` is the type variable from `java.util.List`, which has no meaning outside of `java.util.List`. In general, calling `QualIdent` with a `ClassSymbol`, I don't see how any type other than the erased one would make sense.

That sad - I can easily remove this part of the patch, the AST distortion is fairly small, at least at this time.

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/TreeMaker.java:749]: I see what you mean, then I think we should document in the method that the produced AST will be erased, also please check double check if the type of JCFieldAccess::selected needs to be erased too, I think we should be fine but just to make sure

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

**Pull request closed without merging**